urllib3/_backends/__init__.py

```diff
@@ -1,9 +1,56 @@
+import logging
+
 from urllib3.packages import six
-from .sync_backend import SyncBackend
 
-__all__ = ['SyncBackend']
 
-if six.PY3:
+def load_trio_backend():
+    if not six.PY3:
+        raise ValueError("trio backend requires Python 3")
     from .trio_backend import TrioBackend
+    return TrioBackend
+
+
+def load_twisted_backend():
+    if not six.PY3:
+        raise ValueError("twisted backend requires Python 3")
     from .twisted_backend import TwistedBackend
-    __all__ += ['TrioBackend', 'TwistedBackend']
+    return TwistedBackend
+
+
+def load_sync_backend():
+    from .sync_backend import SyncBackend
+    return SyncBackend
+
+
+BACKENDS = {
+    "sync": load_sync_backend,
+    "trio": load_trio_backend,
+    "twisted": load_twisted_backend,
+}
+
+
+def available_backends():
+    """
+    Returns a list of backend names that we are able to successfully import.
+    """
+    output = []
+    for backend_name, load_backend_fn in sorted(BACKENDS.items()):
+        try:
+            load_backend_fn()
+            output.append(backend_name)
+        except:  # noqa
+            logging.exception("{} not available".format(backend_name))
+    return output
+
+
+def load_backend(backend_name, backend_args):
+    _backend_args = backend_args or {}
+    if backend_name is None:
+        return load_sync_backend()(**_backend_args)
+    elif backend_name in BACKENDS:
+        return BACKENDS[backend_name]()(**_backend_args)
+    else:
+        valid_backends = ", ".join(sorted(BACKENDS.keys()))
+        raise ValueError("Backend must be one of: {}".format(valid_backends))
```

Comment:
Another option would be to use a static list of backends in the `else` clause, and let the loading of a backend fail if the dependency is not installed. Which might be more intuitive?

Reply:
I'm not sure I follow - can you rephrase your suggestion?

This else block is raising an error if the user specifies an unknown backend. At this point, I'm assuming that we will have a finite list of backends in urllib3 and will not allow users to write their own drivers without adding them upstream - maybe we want to reconsider this, though.

Reply:
Sorry, this was not easy to understand. I'll try to be clearer this time.

So I had two suggestions. The first one, above, was simply about replacing `except` by `except Exception` or `except ImportError`.

This suggestion would have more impact. I was wondering about hard coding the valid backends (currently that would be trio, twisted and sync) and always report them as available in the ValueError raised in the else clause of `load_backend`. I think this would help with the discovery of new backends. Now, if a user tries to use such a backend but the import fails, I don't know if we should catch the ImportError exception or let it propagate.

I'm really thinking aloud here: I don't know if that would be better or not!

Reply:
@pquentin Apologies - I think I've misled you. The `available_backends` function is not used - the error I am raising contains all "known" backends.

I added `available_backends` to assist in debugging from a repl.

Reply:
I'm going to remove the `available_backends` function as it serves no purpose beyond debugging, and probably wouldn't be exposed as a user visible API.

Reply:
Not sure how I managed to misread this. Sorry!

I think we can leave available_backends as it is.

Reply:
Sorry for the back and forth on this. But it didn’t make sense for me to implement available_backends just yet without agreeing on some kind of user interface for it. I’d like to do that separately.

We also need to be careful with that functionality - as it will import all of the backends, it’s probably something that we don’t want to encourage users to call.